dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,387 @@
+void NumberBox::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected = *this;
+
+    if (auto spinDown = GetTemplateChildT<winrt::RepeatButton>(L"DownSpinButton", controlProtected))
+    {
+        spinDown.Click({ this, &NumberBox::OnSpinDownClick });
+
+        // Do localization for the down button
+        if (winrt::AutomationProperties::GetName(spinDown).empty())
+        {
+            auto spinDownName = ResourceAccessor::GetLocalizedStringResource(SR_NumberBoxDownSpinButtonName);
+            winrt::AutomationProperties::SetName(spinDown, spinDownName);
+        }
+    }
+
+    if (auto spinUp = GetTemplateChildT<winrt::RepeatButton>(L"UpSpinButton", controlProtected))
+    {
+        spinUp.Click({ this, &NumberBox::OnSpinUpClick });
+
+        // Do localization for the up button
+        if (winrt::AutomationProperties::GetName(spinUp).empty())
+        {
+            auto spinUpName = ResourceAccessor::GetLocalizedStringResource(SR_NumberBoxUpSpinButtonName);
+            winrt::AutomationProperties::SetName(spinUp, spinUpName);
+        }
+    }
+
+    m_textBox.set([this, controlProtected]() {
+        auto textBox = GetTemplateChildT<winrt::TextBox>(L"InputBox", controlProtected);
+        if (textBox)
+        {
+            textBox.LostFocus({ this, &NumberBox::OnTextBoxLostFocus });
+            textBox.KeyUp({ this, &NumberBox::OnNumberBoxKeyUp });
+        }
+        return textBox;
+    }());
+
+    // Initializing precision formatter. This formatter works neutrally to protect against floating point imprecision resulting from stepping/calc
+    m_stepPrecisionFormatter.FractionDigits(0);
+    m_stepPrecisionFormatter.IntegerDigits(1);
+    m_stepPrecisionFormatter.NumberRounder(nullptr);
+    m_stepPrecisionRounder.RoundingAlgorithm(winrt::RoundingAlgorithm::RoundHalfAwayFromZero);
+
+    SetSpinButtonVisualState();
+    UpdateTextToValue();
+}
+
+void NumberBox::OnValuePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    auto oldValue = unbox_value<double>(args.OldValue());
+
+    if (std::isnan(Value()) && BasicValidationMode() == winrt::NumberBoxBasicValidationMode::InvalidInputOverwritten)
+    {
+        // In the validation case, we don't consider NaN to be valid.
+        Value(oldValue);
+    }
+
+    CoerseValue();
+
+    if (Value() != oldValue)
+    {
+        // Fire ValueChanged event
+        auto valueChangedArgs = winrt::make_self<NumberBoxValueChangedEventArgs>(oldValue, Value());
+        m_valueChangedEventSource(*this, *valueChangedArgs);
+
+        // Fire value property change for UIA
+        if (auto peer = winrt::FrameworkElementAutomationPeer::FromElement(*this).as<winrt::NumberBoxAutomationPeer>())
+        {
+            winrt::get_self<NumberBoxAutomationPeer>(peer)->RaiseValueChangedEvent(oldValue, Value());
+        }
+    }
+
+    UpdateTextToValue();
+}
+
+void NumberBox::OnMinimumPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
```

Comment:
>OnMinimumPropertyChanged [](start = 16, length = 24)

I don't see anything that enforces min < max

Reply:
You are not wrong, there's nothing that does that. It's on my TODO list.